### ISSUE TITLE
Package allow list is not respected if only a `.packages` file is passed

### ISSUE THREAD

**Issue by jakemac53** [dart-lang/sdk#42537]
The behavior I expected is that all packages in the allow list would be opted in in this case, since they get the latest language version.

This behavior is not however fully specified, and making this change today would be very breaking. Primarily this is a question of _how_ exactly things break when only a regular `.packages` file is present. Either some packages in the allow list that are not actually migrated yet would be opted in when they shouldn't be, or migrated versions will not be opted in (the current behavior).

An example repro is to check out the [collection](https://github.com/dart-lang/collection) package, and add this file at `lib/test.dart`:

```dart
main() {
  int? x;
  print(x);
}
```

If you run that with `dart lib/test.dart` it works fine.

If you then add an empty `.packages` file at `lib/.packages` and re-run, it will fail saying the package is not opted in.

**Comment by lrhn** [comment]
The behavior is actually specified. From the [language versioning feature specification](https://github.com/dart-lang/language/blob/master/accepted/future-releases/language-versioning/feature-specification.md#:~:text=no%20language%20version):

> If an entry in the `package_config.json` file contains no language version entry (the "languageVersion" entry is optional), or there is no `package_config.json` file available (yet), the compiler defaults the package to the most recent language version supported by the SDK processing the package. 

For a `.packages` file, the current implementation of `package:package_config` provides a configuration with no `languageVersion` in the packages. That *should* make the package use the SDKs language version, because that's what it means if the language version is missing from the JSON file.

If the experiment is enabled for the package, which the allow list should guarantee, that would be Null Safe semantics for all files. 
Not what you want? Then use a `package_config.json` file! (Really, you should not be using `.packages` files for *anything* any more!)

So, this does appear to be a bug.

**Comment by leafpetersen** [comment]
@lrhn is there any danger that users will get silently broken by this in confusing ways if they have an existing project (with a .packages file), download a new SDK (with an allow list), and then try to run without doing something that causes a package_config.json to be run?

**Comment by lrhn** [comment]
It could probably break someone. 

If they have an existing Dart script deployed along with a `.packages` file including a `collection` package, then they might no longer be able to run it just by doing `dart script.dart`.

The same could be said for a Dart script along with a `packages/` directory, and we have broken that already. It's not about *if* the `.packages` setup breaks, but about *when* we break it. If it's too soon, then we might need to do something to avoid it.

Will it break older `pub global activate`ed scripts? (That is, will they have only a `.packages` file?) @jonasfj 

We could choose to change our behavior, so that any code running with just `.packages` is locked to Dart version 2.6 (the language version pre-language versioning). They would not be able to use newer features, but if they were written using those features, they would also be written using an SDK which generates a `.dart_tool/package_config.json` file.

**Comment by jakemac53** [comment]
> You added an _empty_ `.packages` file, and it was found as the package configuration for the script being run.
> That script is not in `package:collection`. It can't be, because there is no packages at all in the configuration being used, so no files are inside the root directory of the `collection` package.

Sorry I was trying to distill this to a simple repro and over-simplified :D. You can similarly reproduce this if the .packages file has an entry for package:collection.

**Comment by jakemac53** [comment]
> If the experiment is enabled for the package, which the allow list should guarantee, that would be Null Safe semantics for all files.

Yes, this was the expected behavior and is not the current behavior. That is what this issue is about :).

**Comment by lrhn** [comment]
Ack, then it does sound like a bug.

**Comment by jonasfj** [comment]
> Will it break older `pub global activate`ed scripts? (That is, will they have only a .packages file?) @jonasfj

Yes, it probably will break these. But mind you we've been generating a `package_config.json` since Dart `2.7.0`.

**Comment by johnniwinther** [comment]
@lrhn I was trying recreate the difference in having a `.packages` file and a `packages_config.json` file without an explicit language for a package. Here the `Package.languageVersion` field for `.packages` is set to a specific version, whereas for `packages_config.json` it would be `null` (as expected). Is it intentional that a `Package` object for a `.packages` file is marked as having an explicit version?

**Comment by johnniwinther** [comment]
It seems that packages derived from `.packages` files are deliberately set to have version `2.7`. See https://github.com/dart-lang/package_config/blob/9c586d04bd26fef01215fd10e7ab96a3050cfa64/lib/src/packages_file.dart#L14

This means that it is expected behavior that we do not enable allowed experiments from `.packages` - their language version is simply too low.

@leafpetersen @jakemac53 Is this acceptable behavior or do we need to support `.packages` with allowed experiments?

**Comment by jakemac53** [comment]
Well, it doesn't implement [the spec](https://github.com/dart-lang/language/blob/master/accepted/future-releases/language-versioning/feature-specification.md), and it diverges in a way that is quite user visible in the scenarios where it does come up.

So I would say that we should change the implementation to return the current language version and not 2.7. This is likely just an oversight?

**Comment by johnniwinther** [comment]
This is not for the `package_config.json` (it does behave as spec'ed) but for `.packages`.

In any case, if this is a bug, it is a `package:package_config` bug. The CFE just queries that package.

**Comment by jakemac53** [comment]
Note this part of the linked text `or there is no package_config.json file available (yet)` - this covers the `.packages` file only context.

I agree it seems to be a package_config issue.
